### DISCUSSION TITLE
How to install large Contract Bundles exceeding the defined `max_tx_bytes`

### DISCUSSION THREAD

**Discussion by Jorge-Lopes** [Agoric/agoric-sdk#9455] [Q&A] [answered]
Hello everyone,

I wish to discuss potential solutions or workarounds for an issue that is triggered  when a developer attempts to install a contract bundle that exceeds the default size limits set for the RPC request bodies. 

This topic was recently addressed on this [office hours](https://github.com/Agoric/agoric-sdk/discussions/9419) session with @dckc.

**Steps to Reproduce:**

1. Develop a contract bundle with a size larger than the default RPC node limit for request bodies (1M)
2. Attempt to install the contract bundle using the command `agd tx swingset install-bundle ...`
3. Observe the error message indicating that the bundle size exceeds the permitted limit `Invalid Request: error reading request body: http: request body too large` .

**Bundle Explorer:** This [tool](https://github.com/Agoric/agoric-sdk/discussions/8416) allow us to visualise the modules present on the bundle and their size. It can be very useful to identify an abnormally large module or duplicated ones.

**Proposed Solutions:**

1. **Test Bundle Size**: see this [issue](https://github.com/Agoric/dapp-offer-up/issues/36) for more detail
2. **Compress Bundle:** while installing the bundle, provide the `--compress` flag
3. **Prevent duplicated packages:** see this [issue](https://github.com/Agoric/agoric-sdk/issues/8621) for more details on `yarn-deduplicate vs resolutions`

**Questions:**

1. Are there any best practices for  handling large contract bundles? Such as splitting the bundle/contract.
2. Is there a recommended on how to modify the RPC node size limits to accommodate larger contract bundles.

Thank you for your time,
Best regards

**Comment by dckc** [comment]
@michaelfig do you remember the fragmentation-and-reassembly hack we did? How do I find that?

**Reply by michaelfig** [reply]
IIRC, `dc-installInPieces` was the branch it was on.

(I got to that by searching through `git branch` for relevant terms.  I wouldn't've found it except it was on the line after `dc-deploy-ocap` and "deploy" was one of the terms I searched for. 😅)

**Reply by dckc** [reply]
bingo! Thanks, @michaelfig 

@Jorge-Lopes the main commit was:

 - **feat(run-protcol): installInPieces for runStake** https://github.com/Agoric/agoric-sdk/commit/daae93494ebf9543be8683064b7863ad790e3cfc
> runStake is too big to E(zoe).install() in agorictest-8
> so we have a helper contract

There were one or two follow-up tweaks to the deploy script and the reassembly contract, resulting in:

https://github.com/Agoric/agoric-sdk/blob/dc-installInPieces/packages/run-protocol/test/runStake/runStake-deploy.js
https://github.com/Agoric/agoric-sdk/blob/dc-installInPieces/packages/run-protocol/test/runStake/endoCAS.js

**Reply by dckc** [reply, marked as answer]
btw... that's a horrible hack. We try to keep large bundles out of messages that get passed between vats.

We track actual solutions in...

 - https://github.com/Agoric/agoric-sdk/issues/2391

Some prototyping showed that 90% of bundle contents don't need to be uploaded again; they're already on chain.
